src/build/buildController.ts

```diff
@@ -0,0 +1,178 @@
+import * as vscode from 'vscode';
+import * as fs from 'fs-extra';
+import * as path from 'path';
+import { Credential } from '../credential/CredentialController';
+import { OPBuildAPIClient } from './OPBuildAPIClient';
+import { TriggerErrorType } from './TriggerErrorType';
+import { EventStream } from '../common/eventStream';
+import { DiagnosticController } from './diagnosticController';
+import { safelyReadJsonFile, getRepositoryInfoFromLocalFolder } from '../utils/utils';
+import { EnvironmentController } from '../common/environmentController';
+import { BuildExecutor } from './buildExecutor';
+import { PlatformInformation } from '../common/platformInformation';
+import { OUTPUT_FOLDER_NAME } from '../shared';
+import { visualizeBuildReport } from './reportGenerator';
+import { BuildJobSucceeded, BuildTriggerFailed, BuildInstantAllocated, BuildInstantReleased, BuildProgress, RepositoryInfoRetrieved, BuildJobTriggered, BuildJobFailed } from '../common/loggingEvents';
+
+export class BuildController {
+    private activeWorkSpaceFolder: vscode.WorkspaceFolder;
+    private instantAvailable: boolean;
+    private repositoryUrl: string;
+    private repositoryBranch: string;
+    private opBuildAPIClient: OPBuildAPIClient;
+    private buildExecutor: BuildExecutor;
+
+    constructor(
+        environmentController: EnvironmentController,
+        platformInformation: PlatformInformation,
+        private diagnosticController: DiagnosticController,
+        private eventStream: EventStream,
+    ) {
+        this.instantAvailable = true;
+
+        this.opBuildAPIClient = new OPBuildAPIClient(environmentController);
+        this.buildExecutor = new BuildExecutor(platformInformation, environmentController, eventStream);
+    }
+
+    public async build(uri: vscode.Uri, credential: Credential): Promise<void> {
+        if (!this.trySetAvailableFlag()) {

```

Comment:
>trySetAvailableFlag [](start = 18, length = 19)

How about use the name: Semaphore and Release?

Reply:
Semaphore is a integer variable? it is just a boolean flag here.

---
In reply to: [365632717](https://github.com/docascode/vscode-docs-build/pull/13#discussion_r365632717) [](ancestors = 365632717)